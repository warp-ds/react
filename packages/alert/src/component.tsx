import React, { PropsWithChildren, ReactElement } from 'react';

import { classNames } from '@chbphone55/classnames';
import { alert as ccAlert } from '@warp-ds/css/component-classes';
import IconError16 from '@warp-ds/icons/react/error-16';
import IconInfo16 from '@warp-ds/icons/react/info-16';
import IconSuccess16 from '@warp-ds/icons/react/success-16';
import IconWarning16 from '@warp-ds/icons/react/warning-16';

import { ExpandTransition } from '../../_helpers/index.js';

import type { AlertProps } from './props.js';

export function Alert({ show, type, role = 'alert', children, ...props }: PropsWithChildren<AlertProps>) {
  const variantClasses = ccAlert[type];
  const iconVariantClasses = ccAlert[`${type}Icon`];

  return (
    <ExpandTransition show={show}>
      <div role={role} className={classNames(props.className, `${ccAlert.alert} ${variantClasses}`)} style={props.style}>
        <div className={`${ccAlert.icon} ${iconVariantClasses}`}>{iconMap[type]}</div>
        <div className={ccAlert.textWrapper}>{children}</div>
      </div>
    </ExpandTransition>
  );
}

const iconMap: {
  // eslint-disable-next-line
  [key in AlertProps['type']]: ReactElement;
} = {
  negative: <IconError16 />,
  positive: <IconSuccess16 />,
  warning: <IconWarning16 />,
  info: <IconInfo16 />,
};
