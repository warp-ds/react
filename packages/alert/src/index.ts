export { Alert } from './component.js';
export type { AlertProps } from './props.js';
