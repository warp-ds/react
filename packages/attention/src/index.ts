export { Attention } from './component.js';
