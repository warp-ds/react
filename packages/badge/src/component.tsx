import { classNames } from '@chbphone55/classnames';
import { badge as ccBadge } from '@warp-ds/css/component-classes';
import React from 'react';
import { BadgeProps } from './props.js';

export function Badge(props: BadgeProps) {
  const {
    children,
    as = 'div',
    variant = 'neutral',
    position,
    ...rest
  } = props;

  return React.createElement(
    as,
    {
      ...(rest as Omit<BadgeProps, 'children'> as {}),
      className: classNames(
        ccBadge.base,
        ccBadge[variant],
        {
            [ccBadge.positionBase]: !!position,
            [ccBadge.positionTL]: position === 'top-left',
            [ccBadge.positionTR]: position === 'top-right',
            [ccBadge.positionBR]: position === 'bottom-right',
            [ccBadge.positionBL]: position === 'bottom-left',
        },
        props.className,
      ),
    },
    children,
  );
}
