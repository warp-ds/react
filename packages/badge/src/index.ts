export { Badge } from './component.js';
export type { BadgeProps } from './props.js';
