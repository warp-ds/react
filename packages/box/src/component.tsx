import { classNames } from '@chbphone55/classnames';
import { box as ccBox } from '@warp-ds/css/component-classes';
import React from 'react';
import { BoxProps } from './props.js';

export function Box(props: BoxProps) {
  const {
    children,
    as = 'div',
    bleed,
    clickable,
    neutral,
    bordered,
    info,
    role,
    ...rest
  } = props;

  return React.createElement(
    as,
    {
      ...(rest as Omit<BoxProps, 'children'> as {}),
      role: role ?? 'region',
      className: classNames(
        ccBox.box,
        {
          [ccBox.bleed]: bleed,
          [ccBox.info]: info,
          [ccBox.neutral]: neutral,
          [ccBox.bordered]: bordered,
          [ccBox.infoClickable]:clickable && info,
          [ccBox.neutralClickable]: clickable && neutral,
          [ccBox.borderedClickable]: clickable && bordered,
        },
        props.className,
      ),
    },
    children,
  );
}
