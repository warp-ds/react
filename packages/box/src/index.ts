export { Box } from './component.js';
export type { BoxProps } from './props.js';
