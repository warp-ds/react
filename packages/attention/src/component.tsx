import React, { forwardRef, useEffect, useMemo, useRef, useState } from 'react';

import { classNames } from '@chbphone55/classnames';
import { i18n } from '@lingui/core';
import { arrowDirectionClassname, opposites, useRecompute as recompute } from '@warp-ds/core/attention';
import { attention as ccAttention } from '@warp-ds/css/component-classes';
import IconClose16 from '@warp-ds/icons/react/close-16';

import { activeAttentionType, getVariant, pointingAtDirection, useAutoUpdatePosition } from '../../_helpers/attention.js';
import { activateI18n } from '../../i18n.js';

import { messages as enMessages } from './locales/en/messages.mjs';
import { messages as fiMessages } from './locales/fi/messages.mjs';
import { messages as nbMessages } from './locales/nb/messages.mjs';
import type { ArrowProps, AttentionProps, ReferenceElement } from './props.js';

const variantClasses = {
  callout: {
    wrapper: ccAttention.callout,
    arrow: ccAttention.arrowCallout,
  },
  highlight: {
    wrapper: ccAttention.highlight,
    arrow: ccAttention.arrowHighlight,
  },
  tooltip: {
    wrapper: ccAttention.tooltip,
    arrow: ccAttention.arrowTooltip,
  },
  popover: {
    wrapper: ccAttention.popover,
    arrow: ccAttention.arrowPopover,
  },
};

export function Attention(props: AttentionProps) {
  let {
    noArrow,
    isShowing,
    children,
    role,
    'aria-label': ariaLabel,
    placement = 'bottom',
    targetEl: initialTargetEl,
    className,
    canClose,
    onDismiss,
    distance = 8,
    skidding = 0,
    flip = false,
    crossAxis = false,
    fallbackPlacements,
    ...rest
  } = props;

  activateI18n(enMessages, nbMessages, fiMessages);

  const wrapperClasses = classNames(ccAttention.base, variantClasses[getVariant(rest, variantClasses)].wrapper);

  const [actualDirection, setActualDirection] = useState(placement);
  // Don't show attention element before its position is computed on first render
  const [isVisible, setIsVisible] = useState<boolean | undefined>(false);

  const isMounted = useRef(true);
  const attentionEl = useRef<HTMLDivElement | null>(null);
  const arrowEl = useRef<HTMLDivElement | null>(null);
  const targetElRef = useRef<ReferenceElement | null>(initialTargetEl?.current || null);

  const attentionState = useMemo(
    () => ({
      get isShowing() {
        return isShowing;
      },
      get isCallout() {
        return rest.callout;
      },
      get actualDirection() {
        return actualDirection;
      },
      set actualDirection(v) {
        setActualDirection(v);
      },
      get directionName() {
        return placement;
      },
      get arrowEl() {
        return arrowEl.current;
      },
      get attentionEl() {
        return attentionEl.current;
      },
      set attentionEl(v) {
        attentionEl.current = v;
      },
      get targetEl() {
        return targetElRef?.current;
      },
      get noArrow() {
        return props.noArrow;
      },
      get distance() {
        return distance;
      },
      get skidding() {
        return skidding;
      },
      get flip() {
        return flip;
      },
      get crossAxis() {
        return crossAxis;
      },
      get fallbackPlacements() {
        return fallbackPlacements;
      },
    }),
    [
      isShowing,
      rest.callout,
      actualDirection,
      placement,
      arrowEl,
      attentionEl,
      targetElRef,
      props.noArrow,
      distance,
      skidding,
      flip,
      crossAxis,
      fallbackPlacements,
    ],
  );

  const defaultAriaLabel = () => `${activeAttentionType(props)} ${!props.noArrow ? pointingAtDirection(actualDirection) : ''}`;

  useEffect(() => {
    // targetEl can be undefined if props.callout is true.
    // However, useAutoUpdatePosition hook is using @warp-ds/core, which uses Floating-ui's computePosition(). Floating-ui's computePosition() requires a defined targetEl to be able to compute the attentionEl's position and the attentionEl's arrow position.
    // When props.callout is true, we only need computePosition() to calculate the callout's arrow position. So, we create a default targetEl for callout that we can pass to the useAutoUpdatePosition hook, in order to avoid Floating-ui from throwing an error.
    if (props.callout && initialTargetEl === undefined) {
      targetElRef.current = document?.createElement('div');
    } else {
      targetElRef.current = initialTargetEl?.current || null;
    }
  }, [props.callout, initialTargetEl]);

  useEffect(() => {
    recompute(attentionState);
  }, [attentionState]);

  useEffect(() => {
    if (isMounted.current) {
      isMounted.current = false;
      // update attention's visibility after first render if showing by default or it's of type callout
      if (isShowing || props.callout) {
        setIsVisible(isShowing);
      }
    } else {
      setIsVisible(isShowing);
    }
  }, [isShowing, props.callout]);

  useAutoUpdatePosition(targetElRef, isShowing, attentionEl, attentionState);

  return !props.callout && initialTargetEl === undefined ? null : (
    <div
      data-testid="attention-el"
      className={classNames(
        {
          [ccAttention.notCallout]: !props.callout,
          invisible: !isVisible && !props.callout,
          hidden: !isVisible && !props.tooltip,
        },
        className,
      )}
      ref={attentionEl}>
      <div
        role={props.role === '' ? undefined : props.tooltip ? 'tooltip' : 'img'}
        aria-label={ariaLabel === '' ? undefined : ariaLabel ?? defaultAriaLabel()}
        className={wrapperClasses}
        id={props.id}>
        {!props.noArrow && <Arrow {...props} ref={arrowEl} direction={actualDirection} />}
        <div className={ccAttention.content}>{props.children}</div>
        {canClose && (
          <button
            type="button"
            aria-label={i18n._(
              /*i18n*/ {
                id: 'attention.aria.close',
                message: 'Close',
                comment: 'Aria label for the close button in attention',
              },
            )}
            onClick={onDismiss}
            onKeyDown={(event) => {
              if (!props.onDismiss) return;
              if (event.key === 'Escape') {
                props.onDismiss();
              }
            }}
            className={ccAttention.closeBtn}>
            <IconClose16 />
          </button>
        )}
      </div>
    </div>
  );
}

const Arrow = forwardRef<HTMLDivElement, ArrowProps>(({ direction, ...rest }, ref) => {
  const arrowDirection = opposites[direction];
  const arrowClasses = classNames(
    ccAttention.arrowBase,
    ccAttention[`arrowDirection${arrowDirectionClassname(arrowDirection)}`],
    variantClasses[getVariant(rest, variantClasses)].arrow,
  );

  return <div data-testid="attention-arrow-el" ref={ref} className={arrowClasses} />;
});
