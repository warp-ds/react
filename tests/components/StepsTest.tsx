import React from 'react';

import { render, screen } from '@testing-library/react';

import { Steps } from '../../packages/steps/src/component';
import { Step } from '../../packages/steps/src/step';

describe('Steps component', () => {
  it('renders a ul element with the correct class names', () => {
    const { container } = render(<Steps horizontal={true} right={false} />);
    const ulElement = container.querySelector('ul');
    expect(ulElement).toHaveClass('w-full flex');
  });

  it('renders a ul element with the correct class names vertical', () => {
    const { container } = render(<Steps horizontal={false} right={false} />);
    const ulElement = container.querySelector('ul');
    expect(ulElement).toHaveClass('w-full');
  });

  it('renders children elements', () => {
    const childElement = <li>Child element</li>;
    const { container } = render(<Steps>{childElement}</Steps>);
    expect(container.querySelector('li')).toBeInTheDocument();
  });
});

describe('Step component', () => {
  it('renders correctly with default props', () => {
    const { container } = render(<Step>Step content</Step>);
    // dom is a bit complicated hence snapshot testing
    expect(container).toMatchSnapshot();
  });

  it('renders correctly with active prop', () => {
    const { container } = render(<Step active>Step content</Step>);
    expect(container).toMatchSnapshot();
  });

  it('renders correctly with completed prop', () => {
    const { container } = render(<Step completed>Step content</Step>);
    expect(container).toMatchSnapshot();
  });

  it('renders correctly with both active and completed props', () => {
    const { container } = render(
      <Step active completed>
        Step content
      </Step>,
    );
    expect(container).toMatchSnapshot();
  });

  it('renders correct aria-label', () => {
    render(<Step active>Step content</Step>);
    const stepDot = screen.getByRole('img');
    expect(stepDot).toHaveAttribute('aria-label', 'Step indicator active circle');
  });

  it('renders correct aria-current attribute', () => {
    render(<Step active>Step content</Step>);
    const stepDot = screen.getByRole('img');
    expect(stepDot).toHaveAttribute('aria-current', 'step');
  });

  it('renders IconCheck16 when completed prop is true', () => {
    render(<Step completed>Step content</Step>);
    const icon = screen.getByTestId('completed-icon');
    expect(icon).toBeInTheDocument();
  });
});
