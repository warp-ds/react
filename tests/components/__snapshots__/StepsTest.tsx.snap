// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Step component > renders correctly with active prop 1`] = `
<div>
  <li
    class="group/stepv grid-rows-[20px_auto] grid grid-flow-col gap-x-16 grid-cols-[20px_1fr]  "
  >
    <div
      aria-current="step"
      aria-label="Step indicator active circle"
      class="rounded-full border-2 h-20 w-20 transition-colors duration-300 s-icon-inverted    s-border-primary s-bg-primary"
      role="img"
    />
    <div
      class="group-last/stepv:hidden transition-colors duration-300 group-last/steph:bg-transparent w-2 h-full justify-self-center   s-bg-disabled "
    />
    <div
      class="last:mb-0 group-last/step:last:pb-0 row-span-2 pb-32 "
    >
      Step content
    </div>
  </li>
</div>
`;

exports[`Step component > renders correctly with both active and completed props 1`] = `
<div>
  <li
    class="group/stepv grid-rows-[20px_auto] grid grid-flow-col gap-x-16 grid-cols-[20px_1fr]  "
  >
    <div
      aria-current="step"
      aria-label="Step indicator completed circle"
      class="rounded-full border-2 h-20 w-20 transition-colors duration-300 s-icon-inverted    s-border-primary s-bg-primary"
      role="img"
    >
      <svg
        data-testid="completed-icon"
        fill="none"
        height="16"
        viewBox="0 0 16 16"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <title>
          Checkmark
        </title>
        <path
          d="M3.83 7.667 7.163 11l5.334-6"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="1.5"
        />
      </svg>
    </div>
    <div
      class="group-last/stepv:hidden transition-colors duration-300 group-last/steph:bg-transparent w-2 h-full justify-self-center    s-bg-primary"
    />
    <div
      class="last:mb-0 group-last/step:last:pb-0 row-span-2 pb-32 "
    >
      Step content
    </div>
  </li>
</div>
`;

exports[`Step component > renders correctly with completed prop 1`] = `
<div>
  <li
    class="group/stepv grid-rows-[20px_auto] grid grid-flow-col gap-x-16 grid-cols-[20px_1fr]  "
  >
    <div
      aria-label="Step indicator completed circle"
      class="rounded-full border-2 h-20 w-20 transition-colors duration-300 s-icon-inverted    s-border-primary s-bg-primary"
      role="img"
    >
      <svg
        data-testid="completed-icon"
        fill="none"
        height="16"
        viewBox="0 0 16 16"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <title>
          Checkmark
        </title>
        <path
          d="M3.83 7.667 7.163 11l5.334-6"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="1.5"
        />
      </svg>
    </div>
    <div
      class="group-last/stepv:hidden transition-colors duration-300 group-last/steph:bg-transparent w-2 h-full justify-self-center    s-bg-primary"
    />
    <div
      class="last:mb-0 group-last/step:last:pb-0 row-span-2 pb-32 "
    >
      Step content
    </div>
  </li>
</div>
`;

exports[`Step component > renders correctly with default props 1`] = `
<div>
  <li
    class="group/stepv grid-rows-[20px_auto] grid grid-flow-col gap-x-16 grid-cols-[20px_1fr]  "
  >
    <div
      aria-label="Empty circle"
      class="rounded-full border-2 h-20 w-20 transition-colors duration-300 s-icon-inverted   s-border s-bg "
      role="img"
    />
    <div
      class="group-last/stepv:hidden transition-colors duration-300 group-last/steph:bg-transparent w-2 h-full justify-self-center   s-bg-disabled "
    />
    <div
      class="last:mb-0 group-last/step:last:pb-0 row-span-2 pb-32 "
    >
      Step content
    </div>
  </li>
</div>
`;
